buildtools/dep/dep_test.go

```diff
@@ -0,0 +1,80 @@
+package dep_test
+
+import (
+	"testing"
+
+	"github.com/fossas/fossa-cli/buildtools/dep"
+	"github.com/fossas/fossa-cli/pkg"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestIsIgnored(t *testing.T) {
+	manifest := dep.Manifest{Ignored: []string{"apple", "orange*"}}
+
+	// File listed in ignored list is ignored.
+	valid := manifest.IsIgnored("apple")
+	assert.Equal(t, valid, true)
+
+	// Wildcard entry properly ignores its own package.
+	valid = manifest.IsIgnored("orange")
+	assert.Equal(t, valid, true)
+
+	// Wildcard entry properly ignores other packages.
+	valid = manifest.IsIgnored("orange/blood")
+	assert.Equal(t, valid, true)
+
+	// File not listed in ignored list is not ignored.
+	valid = manifest.IsIgnored("apple/fuji")
+	assert.Equal(t, valid, false)
+}
+
+func TestReadLockfile(t *testing.T) {
+	// Reading a valid lockfile returns expected data.
+	lockfile, err := dep.ReadLockfile("testdata/Gopkg.lock")
+	expectedLockfile := dep.Lockfile{
+		Projects: []dep.Project{
+			dep.Project{
+				Name:     "cat/fossa",
+				Packages: []string{"."},
+				Revision: "1",
+				Version:  "v0.3.0",
+			},
+		},
+		Normalized: map[string]pkg.Import{
```

Comment:
Ah, so this is why `Normalized` is public.

I'm not a big fan of this kind of test. We're testing that the `Lockfile`'s _structure_ is what we expect, but the contract that a `Lockfile` presents to its consumers is its behaviour (i.e. that `Resolve` will work correctly).

In this case, this would be a good reason for `Lockfile` to have its own implementation of `Resolve`: so we can test that `Resolve` works as expected rather than asserting against the internal structure of the lockfile.

Alternatively, we can make `lockfile`s themselves private. In that case, they expose no public contract so it's perfectly fine to write an internal test (i.e. one in `package dep`) to check that its internals work.

In general, _external_ tests (i.e. those in `package foo_test`) should ensure that contracts presented to consumers are upheld while _internal_ tests (i.e. those in `package foo`) check that internal data structures and algorithms are working correctly and tend to change more frequently as the internals of the package change.

Reply:
I like the idea of making `Lockfile` private as well as `ReadLockfile`, `ReadManifest`, and `IsIgnored`. I think this clarifies how the file should be tested. Externally: `Resolve` should simply return the revision you are looking for or alert that it is ignored or missing. Internally: `IsIgnored`, `ReadLockfile` and `ReadManifest` can then be tested. 

I have not worked with internal go tests and your comment does a really great job of explaining the different approaches. I think splitting the test files clarifies the intention of each test.

Reply:
On the other hand, `Lockfile` is public in every other file, as well as most of its fields.

Reply:
Feel free to fix that in a follow-on PR. I believe that this is currently the case because `Lockfile` _is_ the `golang.Resolver` for the other Go tools.